### ISSUE TITLE
Error: diffs didn't match during apply

### ISSUE THREAD

**Issue by fromonesrc** [hashicorp/terraform#2501]
Getting this during a `tf apply`: ` diffs didn't match during apply. This is a bug with Terraform and should be reported.`

```
$ tf -v
Terraform v0.5.3
```

```
Error applying plan:

19 error(s) occurred:

* 1 error(s) occurred:

* 1 error(s) occurred:

* Error creating Security Group: InvalidGroup.Duplicate: The security group 'stage-[REDACTED]-mmchd-memcached' already exists for VPC 'vpc-b5d9d6d0'
    status code: 400, request id: []
* 1 error(s) occurred:

* 1 error(s) occurred:

* Error creating CacheSubnetGroup: CacheSubnetGroupAlreadyExists: Cache subnet group stage-[REDACTED]-redis already exists.
    status code: 400, request id: [cb2f39b4-1b88-11e5-a8ac-e799ba3d3403]
* 1 error(s) occurred:

* 1 error(s) occurred:

* Error creating CacheSubnetGroup: CacheSubnetGroupAlreadyExists: Cache subnet group stage-[REDACTED]-redis already exists.
    status code: 400, request id: [cb2f60c5-1b88-11e5-a8ac-e799ba3d3403]
* 1 error(s) occurred:

* 1 error(s) occurred:

* Error creating CacheSubnetGroup: CacheSubnetGroupAlreadyExists: Cache subnet group stage-[REDACTED]-redis already exists.
    status code: 400, request id: [cb304b26-1b88-11e5-a8ac-e799ba3d3403]
* 1 error(s) occurred:

* 1 error(s) occurred:

* Error creating CacheSubnetGroup: CacheSubnetGroupAlreadyExists: Cache subnet group stage-[REDACTED]-memcached already exists.
    status code: 400, request id: [cb2f6091-1b88-11e5-aaab-63850f3e3bca]
* 1 error(s) occurred:

* 1 error(s) occurred:

* Error creating CacheSubnetGroup: CacheSubnetGroupAlreadyExists: Cache subnet group stage-[REDACTED]-redis already exists.
    status code: 400, request id: [cb2f3980-1b88-11e5-aaab-63850f3e3bca]
* 1 error(s) occurred:

* 1 error(s) occurred:

* Error creating CacheSubnetGroup: CacheSubnetGroupAlreadyExists: Cache subnet group stage-[REDACTED]-memcached already exists.
    status code: 400, request id: [cb318373-1b88-11e5-aaab-63850f3e3bca]
* 1 error(s) occurred:

* 1 error(s) occurred:

* Error creating CacheSubnetGroup: CacheSubnetGroupAlreadyExists: Cache subnet group stage-[REDACTED]-mmchd already exists.
    status code: 400, request id: [cb304af2-1b88-11e5-aaab-63850f3e3bca]
* 1 error(s) occurred:

* 1 error(s) occurred:

* Error creating CacheSubnetGroup: CacheSubnetGroupAlreadyExists: Cache subnet group stage-[REDACTED]-redis already exists.
    status code: 400, request id: [cb355404-1b88-11e5-aaab-63850f3e3bca]
* 1 error(s) occurred:

* 1 error(s) occurred:

* Error creating CacheSubnetGroup: CacheSubnetGroupAlreadyExists: Cache subnet group stage-[REDACTED]-redis already exists.
    status code: 400, request id: [cb352d29-1b88-11e5-a8ac-e799ba3d3403]
* 1 error(s) occurred:

* aws_launch_configuration.[REDACTED]: diffs didn't match during apply. This is a bug with Terraform and should be reported.
* 1 error(s) occurred:

* aws_launch_configuration.[REDACTED]: diffs didn't match during apply. This is a bug with Terraform and should be reported.
* 1 error(s) occurred:

* aws_launch_configuration.[REDACTED]: diffs didn't match during apply. This is a bug with Terraform and should be reported.
* 1 error(s) occurred:

* aws_launch_configuration.[REDACTED]: diffs didn't match during apply. This is a bug with Terraform and should be reported.
* 1 error(s) occurred:

* aws_launch_configuration.[REDACTED]: diffs didn't match during apply. This is a bug with Terraform and should be reported.
* 1 error(s) occurred:

* aws_launch_configuration.[REDACTED]: diffs didn't match during apply. This is a bug with Terraform and should be reported.
* 1 error(s) occurred:

* aws_launch_configuration.[REDACTED]: diffs didn't match during apply. This is a bug with Terraform and should be reported.
* 1 error(s) occurred:

* aws_launch_configuration.[REDACTED]: diffs didn't match during apply. This is a bug with Terraform and should be reported.
* 1 error(s) occurred:

* aws_launch_configuration.[REDACTED]: diffs didn't match during apply. This is a bug with Terraform and should be reported.

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.

```

**Comment by phinze** [comment]
Hi @fromonesrc - thanks for the report.

It looks like there were a few name collisions in earlier resources that caused the rest of the apply to end up in a bad state.

 - Can you reproduce this error with `TF_LOG=1` set? The debug logs will show a "reason" for the diff mismatch just before the error.
 - Can you share a config snippet that causes the issue for you?

**Comment by chadgrant** [comment]
I just ran into this as well with latest version, TF_LOG=1 didn't seem to show a reason :(

**Comment by fromonesrc** [comment]
@chadgrant can you share the security group resource that is being modified? You may have two ingress rules on same port/protocol with cidr_blocks in one and security_groups in the other.

**Comment by chadgrant** [comment]
@fromonesrc yes, I do. But can't change it because it would cause cyle (circular dependencies)

so some of the same rules with different sgs are added with aws_security_group_rule to avoid the circular dependency.

**Comment by mmell** [comment]
I saw this error from a `aws_iam_server_certificate` with a lifecycle block:
```
  lifecycle {
    prevent_destroy = true
    ignore_changes = [
      "name",
      "certificate_body",
      "certificate_chain",
      "path",
      "private_key"
    ]
  }
```
I removed the lifecycle block and the problem disappeared.

**Comment by dsolsona** [comment]
I saw this error from a `aws_instance` with the following block
```  
lifecycle {
  ignore_changes = ["user_data"]
}
```

Removing it fixes the issue.

**Comment by phinze** [comment]
Hi folks, if any of you are running to this in Terraform v0.6.12, can you file a fresh issue w/ repro steps? We addressed the `ignore_changes` bug in v0.6.12 so I want to separate out any remaining issues. Thanks!

**Comment by greg-jaunt** [comment]
I'm getting this in 0.6.14 with:

lifecycle {
    ignore_changes = ["tags.LaunchedBy"]
}

I'm passing in the value that this is originally set to as a TF_VAR, so it's expected to change from run-to-run.

I will try to see if there is a simple repo.
